src/v/iceberg/table_update_applier.cc

```diff
@@ -27,10 +27,7 @@ struct update_applying_visitor {
 
     outcome operator()(const add_schema& update) {
         auto sid = update.schema.schema_id;
-        auto s = std::find_if(
-          meta.schemas.begin(),
-          meta.schemas.end(),
-          [sid](const schema& schema) { return schema.schema_id == sid; });
+        auto s = std::ranges::find(meta.schemas, sid, &schema::schema_id);
```

Comment:
@abhijat 

> &schema::schema_id

is schema_id a named type? i'm wondering if this works because named_type has operator() and can be invoked by std::ranges::find. or does std::ranges::find know how to evaluate that member?

Reply:
It's a named type of type `schema::id_t`. `schema_id` itself is a member of the schema.

Reply:
>if it is invoking operator() then this is kind of an anti-pattern because we don't want to compare named_type and primitive type.

Ah it does look like you can pass in an int to find, e.g. I dont think it complains about

```
std::ranges::find(meta.schemas, 4, &schema::schema_id);
```

But you can't pass a different named type. Good callout, it is a little disheartening. The code is much cleaner with it IMO.

Reply:
>std::ranges::find(meta.schemas, 4, &schema::schema_id);
this isn't what i'm talking about. what i'm talking about is a comparison like this

`named_type() == integer`

Reply:
@andijcr did we drop the PR that fixes named_type comparisons to make them stronger? i think if we merged that this code might break

Reply:
> is schema_id a named type? i'm wondering if this works because named_type has operator() and can be invoked by std::ranges::find. or does std::ranges::find know how to evaluate that member?

From what I could see earlier, it is supposed to be able to both access a struct data member and call a member function, eg from the example on [cppreference](https://en.cppreference.com/w/cpp/algorithm/ranges/find):

```
struct folk_info
    {
        unsigned uid;
        std::string name, position;
    };
 
    std::vector<folk_info> folks
    {
        {0, "Ana", "dev"},
        {1, "Bob", "devops"},
        {2, "Eve", "ops"}
    };
 
    const auto who{"Eve"};
    if (auto it = std::ranges::find(folks, who, &folk_info::name); it != folks.end())...
```

Internally the range algorithms use `std::invoke` for projections, which should be able to handle both member functions and struct member data fields (it seems to use `std::is_function` trait to decide what to do).

```
template <class _Ip, class _Sp, class _Pred, class _Proj>
_LIBCPP_HIDE_FROM_ABI static constexpr
_Ip __find_if_impl(_Ip __first, _Sp __last, _Pred& __pred, _Proj& __proj) {
  for (; __first != __last; ++__first) {
    if (std::invoke(__pred, std::invoke(__proj, *__first)))
      break;
  }
  return __first;
}
```

Reply:
I guess we don't control the order in which `std::invoke` checks if a named type is a function or a data member, so there is no way to guarantee that the member will be used for projection and not `operator()`

Reply:
> From what I could see earlier, it is supposed to be able to both access a struct data member and call a member function, eg from the example on [cppreference](https://en.cppreference.com/w/cpp/algorithm/ranges/find):

thanks @abhijat i saw this in the example too it's pretty neat! i find it odd that i need to repeat the type. but oh well!

Reply:
> @andijcr did we drop the PR that fixes named_type comparisons to make them stronger? i think if we merged that this code might break

I'm late to the party, so apologies if I don't get the full context, but I think this code is fine with [that pr](https://github.com/redpanda-data/redpanda/pull/15170). I'll revive it next week, probably.